lib/scope.js

```diff
@@ -883,6 +992,83 @@ module.exports = {
     return buffer.toString( `base64` );
   },
 
+  getSoughtVarPatterns: async function ( scope, { field_name }) {
+    /** Given a var returns data about its use of proxies. If it uses
+    *  proxies (x, i, j, etc.), get all the necessary parts.
+    *  Note: 'proxies' is my own name for this behavior.
+    * 
+    * Example:
+    * x[i].name.first ->
+    * {
+    *   uses_proxies: true,
+    *   proxies: [ 'x', 'i' ],
+    *   names_regex: /^(.+)\[(.+)\]\.name\.first$/,
+    *   name_parts: [ '', 'x', '[', 'i', '].name.first' ]
+    * }
+    */
+    let pattern_data = {
+      uses_proxies: false,  // x, i, j, etc.
+      proxies: [],
+      // TODO: Discuss: Valid regex that won't find a match?
+      names_regex: new RegExp( `this regex should not match anything` ),
+      name_parts: [ field_name ],  // Entire name by default
+    };
+
+    if ( proxies_regex.test( field_name )) {
+      pattern_data.uses_proxies = true;
+      /* substitute, placeholder, proxy, envoy, mediary, wildcard, gap var, catcher, ~catch all~, reserve, generic, space */
+      pattern_data.proxies = field_name.match( proxies_regex );
+      let betweens = field_name.split( proxies_regex )
+
+      // The regex that will discover and capture the actual var names
+      let escaped = [];
+      for ( let tween of betweens ) { escaped.push(
+        tween.replace(/[\-\[\]{}()*+?.,\\\^$|#\s]/g, "\\$&")
+      );}
+      let full_regex = '^' + escaped.join( '(.+)' ) + '$';
+      pattern_data.names_regex = new RegExp( full_regex );
+
+      // A list where every item of the proxy string is included in order
+      // There must be a regex way to do this, but I haven't found it
+      let total_len = pattern_data.proxies.length + betweens.length;
+      let rebuilt = [];
+      for ( let indx = 0; indx < total_len; indx++ ) {
+        if ( indx % 2 === 0 ) {
+          rebuilt.push( betweens[ indx/2 ] );
+        } else {
+          rebuilt.push( pattern_data.proxies[ Math.floor( indx/2 ) ]);
+        }
+      }  // end for rebuilding
+      pattern_data.name_parts = rebuilt;
+    }  // ends if has proxy parts (x, i, j, etc.)
+
+    return pattern_data;
+  },  // Ends scope.getSoughtVarPatterns()
```

Comment:
Are we planning (hypothetically, at some point in the future) to write unit tests for the different functions in this file? This seems like a prime candidate for a lot of unit testing.

Reply:
That question has been on my mind a lot too. I made #197 for it.